File: Java/interview_35.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/
class Solution {
    private Map<Node, Node> map;

    public Node copyRandomList(Node head) {
        if (head == null) {
            return null;
        }
        map = new HashMap<>();
        Node newHead = new Node(head.val);
        Node node = newHead;
        map.put(head, newHead);
        while (head != null) {
            if (head.next == null) {
                node.next = null;
            }
            else if (map.containsKey(head.next)) {
                node.next = map.get(head.next);
            }
            else {
                Node nextNode = new Node(head.next.val);
                node.next = nextNode;
                map.put(head.next, nextNode);
            }
            if (head.random == null) {
                node.random = null;
            }
            else if (map.containsKey(head.random)) {
                node.random = map.get(head.random);
            }
            else {
                Node randomNode = new Node(head.random.val);
                node.random = randomNode;
                map.put(head.random, randomNode);
            }
            head = head.next;
            node = node.next;
        }
        return newHead;
    }
}

File: python/interview_35.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""
class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        ori2copy = {}
        if not head:
            return None
        newHead = Node(head.val)
        ori2copy[head] = newHead
        node = newHead
        while head:
            if head.next and head.next in ori2copy:
                node.next = ori2copy[head.next]
            elif head.next:
                nextNode = Node(head.next.val)
                ori2copy[head.next] = nextNode
                node.next = nextNode
            if head.random and head.random in ori2copy:
                node.random = ori2copy[head.random]
            elif head.random:
                randomNode = Node(head.random.val)
                ori2copy[head.random] = randomNode
                node.random = randomNode
            head = head.next
            node = node.next
        return newHead
